presto-router/src/test/java/com/facebook/presto/router/TestSelectors.java

```diff
@@ -0,0 +1,210 @@
+import com.facebook.airlift.log.Logging;
+import com.facebook.airlift.node.testing.TestingNodeModule;
+import com.facebook.presto.router.cluster.ClusterManager;
+import com.facebook.presto.router.cluster.RequestInfo;
+import com.facebook.presto.router.scheduler.SchedulerType;
+import com.facebook.presto.router.spec.GroupSpec;
+import com.facebook.presto.router.spec.RouterSpec;
+import com.facebook.presto.router.spec.SelectorRuleSpec;
+import com.facebook.presto.server.MockHttpServletRequest;
+import com.facebook.presto.server.security.ServerSecurityModule;
+import com.facebook.presto.server.testing.TestingPrestoServer;
+import com.facebook.presto.tpch.TpchPlugin;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Injector;
+import com.google.inject.Module;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CLIENT_TAGS;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_SOURCE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_USER;
+import static com.facebook.presto.router.scheduler.SchedulerType.RANDOM_CHOICE;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.fail;
+
+public class TestSelectors
+{
+    private List<TestingPrestoServer> prestoServers;
+    private ClusterManager clusterManager;
+    private LifeCycleManager lifeCycleManager;
+
+    @BeforeClass
+    public void setup()
+            throws Exception
+    {
+        File configFile = File.createTempFile("router", "json");
+
+        Logging.initialize();
+
+        // set up server
+        ImmutableList.Builder<TestingPrestoServer> builder = ImmutableList.builder();
+        for (int i = 0; i < 3; ++i) {
+            TestingPrestoServer server = new TestingPrestoServer();
+            server.installPlugin(new TpchPlugin());
+            server.createCatalog("tpch", "tpch");
+            server.refreshNodes();
+            builder.add(server);
+        }
+
+        prestoServers = builder.build();
+
+        List<GroupSpec> groups = new ArrayList<>();
+        List<SelectorRuleSpec> selectors = new ArrayList<>();
+        Optional<SchedulerType> schedulerType = Optional.of(RANDOM_CHOICE);
+        Optional<URI> predictorUri = Optional.empty();
+
+        int i = 1;
+        for (TestingPrestoServer s : prestoServers) {
```

Comment:
nit: It's probably less error prone to use a traditional `for` and just access the servers via index
```suggestion
        for (int i = 0; i < servers.size(); i++) {
```